tests/contrib/patch.py

```diff
@@ -51,151 +56,305 @@ def assert_not_double_wrapped(self, obj):
-            self.assert_wrapped(redis.StrictRedis.pipeline)
-            self.assert_wrapped(redis.Redis.pipeline)
-            self.assert_wrapped(redis.client.BasePipeline.execute)
-            self.assert_wrapped(redis.client.BasePipeline.immediate_execute_command)
-        """
-        raise NotImplementedError(self.test_patch_before_import.__doc__)
+def require_modules(f):
+    """
+    A helper for PatchTestCase test methods that will check if there are any
+    modules to use else raise a NotImplementedError.
 
-    def test_patch_after_import(self):
-        """
-        The integration should test that each class, method or function that
-        is to be patched is in fact done so when ddtrace.patch() is called
-        after the module is imported.
-
-        For example:
-
-        The redis integration patches the following methods:
-        - redis.StrictRedis.execute_command
-        - redis.StrictRedis.pipeline
-        - redis.Redis.pipeline
-        - redis.client.BasePipeline.execute
-        - redis.client.BasePipeline.immediate_execute_command
-
-        an appropriate ``test_patch_after_import`` would be::
-
-            import redis
-            ddtrace.patch(redis=True)
-            self.assert_wrapped(redis.StrictRedis.execute_command)
-            self.assert_wrapped(redis.StrictRedis.pipeline)
-            self.assert_wrapped(redis.Redis.pipeline)
-            self.assert_wrapped(redis.client.BasePipeline.execute)
-            self.assert_wrapped(redis.client.BasePipeline.immediate_execute_command)
-        """
-        raise NotImplementedError(self.test_patch_after_import.__doc__)
+    :param f: method to wrap with a check
+    """
+    def checked_method(self, *args, **kwargs):
+        if not self.MODULES:
+            raise NotImplementedError(f.__doc__)
+        return f(self, *args, **kwargs)
+    return checked_method
 
-    def test_patch_idempotent(self):
-        """
-        Proper testing should be done to ensure that multiple calls to the
-        integration.patch() method are idempotent. That is, that the
-        integration does not patch its library more than once.
-
-        An example for what this might look like is again for the redis
-        integration::
-            ddtrace.contrib.redis.patch()
-            ddtrace.contrib.redis.patch()
-            self.assert_not_double_wrapped(redis.StrictRedis.execute_command)
-        """
-        raise NotImplementedError(self.test_patch_idempotent.__doc__)
 
-    def test_unpatch_before_import(self):
-        """
-        To ensure that we can thoroughly test the installation/patching of an
-        integration we must be able to unpatch it before importing the library.
+class PatchTestCase(object):
+    """
+    unittest or other test runners will pick up the base test case as a testcase
+    since it inherits from unittest.TestCase unless we wrap it with this empty
+    parent class.
+    """
+    @run_in_subprocess
```

Comment:
I make the assumption that all patch test cases should be run in their own subprocess.